enhancements/kube-apiserver/stability-api-removal-notifications.md

```diff
@@ -0,0 +1,159 @@
+---
+title: api-removal-notifications
+authors:
+  - "@sanchezl"
+reviewers:
+  - TBD
+  - "@deads2k"
+approvers:
+  - TBD
+creation-date: 2021-02-01
+last-updated: 2021-02-01
+status: provisional
+---
+
+# API Removal Notifications
+
+> This is the title of the enhancement. Keep it simple and descriptive. A good title can help communicate what the enhancement is and should be considered as part of any review.
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Notify customers that an API that will be removed in the next release is in use.
+
+## Motivation
+
+Kubernetes is being more aggressive with regards to the removal of APIs that have been deprecated or pre-release APIs that have reached a dead-end and will never reach GA. Additionally, the ability to re-enable a previously removed API has been removed.
+
+Customers should be notified that an API that will be removed in the next release is in use so that they can take action to avoid failure when upgrading to the next release. 
+
+### Goals
+
+1. Notify customers an API that will be removed in the next release is in use.
+2. Document how to identify the workload that is using the API to be removed.
+
```

Comment:
I would expect a goal for EUS here.

Reply:
What about filtering workloads by Redhat-provided and user workloads. Can we use APIs in a EUS release in our component that will be gone in EUS+1 ? I would expect this either in goals or in non-goals.

Reply:
Updated Goal/Non-Goals.
> What about filtering workloads by Redhat-provided and user workloads.
This cannot currently be done if relying on alerts for notifications due to the lack of client information in the metrics. Maybe something for a followup enhancement.

Reply:
> What about filtering workloads by Redhat-provided and user workloads.

how do you know the difference?

Reply:
> how do you know the difference?

That's the $100,000 question to make these alert usable.

Reply:
I don't think it matters. We're responsible for cleaning up our mess in the platform, ideally with CI jobs failing on alerts firing. They're responsible for cleaning up their workloads mess, as long as we make it clear which namespace the offending api use is present in is that good enough? I say this assuming that this is knowable with very little knowledge of how this is to be implemented.

Perhaps we also make this a criteria for being in the Red Hat Marketplace, no OLM operators get approved if they're using deprecated APIs.

Reply:
What about resources that don't exist in GA version in EUS1, but do in EUS2, with the beta being removed somewhere in the middle? This is a real danger and I would not call this "mess", but rather something we are in full control to fix during the EUS upgrade path.

I would expect this either be a goal or non-goal to hide these. Making it a non-goal means that we have lots of ugly and the alert/info is always firing.

Reply:
+1 to hiding alerts about OCP-core consumers, however we need to do that.  We'd only want to do that in production though.  Not bothering an admin about something they can't fix outside of updating is great, but forgetting to update our own components and having an API removed that we were still consuming would be really bad.  But before we get too deep into this... how are we planning on distinguishing clients without exploding Prom cardinality?  Would we just filter out particular requests in the metrics-generator and just never submit deprecated-consumption-by-OCP-core-clients to Prom at all?

Reply:
This is still open